File: forge-core/internal/authenticatedadrapprovalcontract/authority.go
package authenticatedadrapprovalcontract

import (
	"fmt"
)

var keyUsages = []string{
	"approval_authorization_state_sign",
	"approval_policy_sign",
	"approval_request_auth",
	"approval_revocation_sign",
	"architecture_approval_sign",
}

var signatureProfileConstants = map[string]string{
	"algorithm":           "Ed25519",
	"api_version":         signatureProfileAPI,
	"canonicalization":    canonicalization,
	"digest_algorithm":    "SHA-256",
	"kind":                "SignatureProfile",
	"message_preimage":    "domain_separator_utf8_nul_then_raw_32_byte_sha256_digest",
	"profile_id":          signatureProfileID,
	"public_key_encoding": "base64url_unpadded_32_bytes",
	"signature_encoding":  "base64url_unpadded_64_bytes",
}

func signatureProfileDocument() map[string]any {
	node := make(map[string]any, len(signatureProfileConstants)+1)
	for field, value := range signatureProfileConstants {
		node[field] = value
	}
	node["profile_sha256"] = signatureProfileSHA256Pin
	return node
}

func validateSignatureProfile(value any) (map[string]any, error) {
	fields := []string{"algorithm", "api_version", "canonicalization", "digest_algorithm",
		"kind", "message_preimage", "profile_id", "profile_sha256",
		"public_key_encoding", "signature_encoding"}
	node, err := requireKeys(value, "SignatureProfile", fields...)
	if err != nil {
		return nil, err
	}
	if _, err = boundedCanonicalJSON(node, maxProfileBytes, "SignatureProfile"); err != nil {
		return nil, err
	}
	for field, expected := range signatureProfileConstants {
		if node[field] != expected {
			return nil, fmt.Errorf("SignatureProfile.%s drifted from v1", field)
		}
	}
	digest, err := signatureProfileSHA256(node)
	if err != nil || node["profile_sha256"] != digest || digest != signatureProfileSHA256Pin {
		return nil, fmt.Errorf("SignatureProfile self digest does not match")
	}
	return node, nil
}

func validateTrustRoot(value any, profileHash string) (map[string]any, error) {
	label := "ArchitectureDecisionApprovalTrustRoot"
	fields := []string{"api_version", "canonicalization", "keys", "kind", "profile_id",
		"root_sha256", "signature_profile_sha256", "trust_domain", "trust_epoch"}
	node, err := requireKeys(value, label, fields...)
	if err != nil {
		return nil, err
	}
	if _, err = boundedCanonicalJSON(node, maxRootBytes, label); err != nil {
		return nil, err
	}
	if node["api_version"] != trustRootAPI || node["canonicalization"] != canonicalization ||
		node["kind"] != label || node["profile_id"] != profileID {
		return nil, fmt.Errorf("%s envelope drifted from v1", label)
	}
	if _, err = textValue(node["trust_domain"], label+".trust_domain", 160); err != nil {
		return nil, err
	}
	if _, err = intValue(node["trust_epoch"], label+".trust_epoch", 1, maxInt64); err != nil {
		return nil, err
	}
	if node["signature_profile_sha256"] != profileHash {
		return nil, fmt.Errorf("%s does not bind the SignatureProfile", label)
	}
	if err = validateRootKeys(node["keys"]); err != nil {
		return nil, err
	}
	digest, err := trustRootSHA256(node)
	if err != nil || node["root_sha256"] != digest {
		return nil, fmt.Errorf("%s self digest does not match", label)
	}
	return node, nil
}

func validateRootKeys(value any) error {
	items, err := sortedUniqueNodes(value, "trust_root.keys", 6, 20)
	if err != nil {
		return err
	}
	keys := make([]map[string]any, len(items))
	for index, item := range items {
		keys[index], err = validateRootKey(item, fmt.Sprintf("trust_root.keys[%d]", index))
		if err != nil {
			return err
		}
	}
	if err = validateUsageCounts(keys); err != nil {
		return err
	}
	return validateKeyDistinctness(keys)
}

func validateRootKey(value any, label string) (map[string]any, error) {
	node, err := requireKeys(value, label, "key_id", "principal", "public_key_base64url", "usage")
	if err != nil {
		return nil, err
	}
	if _, err = textValue(node["key_id"], label+".key_id", 160); err != nil {
		return nil, err
	}
	usage, err := enumValue(node["usage"], label+".usage", keyUsages...)
	if err != nil {
		return nil, err
	}
	allowed := []string{"service"}
	if usage == "architecture_approval_sign" {
		allowed = []string{"human", "operator"}
	} else if usage == "approval_request_auth" {
		allowed = []string{"agent", "human", "operator", "service"}
	}
	if _, err = validatePrincipal(node["principal"], label+".principal", allowed...); err != nil {
		return nil, err
	}
	if _, err = fixedBase64URL(node["public_key_base64url"], label+".public_key_base64url", 32); err != nil {
		return nil, err
	}
	return node, nil
}

func validateUsageCounts(keys []map[string]any) error {
	counts := make(map[string]int)
	for _, key := range keys {
		counts[key["usage"].(string)]++
	}
	for _, usage := range []string{"approval_authorization_state_sign", "approval_policy_sign",
		"approval_request_auth", "approval_revocation_sign"} {
		if counts[usage] != 1 {
			return fmt.Errorf("trust root service/request key usage counts drifted from v1")
		}
	}
	if counts["architecture_approval_sign"] < 2 || counts["architecture_approval_sign"] > 16 {
		return fmt.Errorf("trust root requires 2..16 architecture approval keys")
	}
	return nil
}

func validateKeyDistinctness(keys []map[string]any) error {
	ids, principals, publicKeys := map[string]bool{}, map[string]bool{}, map[string]bool{}
	for _, key := range keys {
		principal, err := canonicalJSON(key["principal"])
		if err != nil {
			return err
		}
		id, publicKey := key["key_id"].(string), key["public_key_base64url"].(string)
		if ids[id] || principals[string(principal)] || publicKeys[publicKey] {
			return fmt.Errorf("root key IDs, principals, and public keys must be pairwise distinct")
		}
		ids[id], principals[string(principal)], publicKeys[publicKey] = true, true, true
	}
	return nil
}

func keyNodesForUsage(root map[string]any, usage string) ([]map[string]any, error) {
	if !containsString(keyUsages, usage) {
		return nil, fmt.Errorf("unsupported trust-root key usage %q", usage)
	}
	var result []map[string]any
	for _, item := range root["keys"].([]any) {
		key := item.(map[string]any)
		if key["usage"] == usage {
			result = append(result, key)
		}
	}
	return result, nil
}

func keyNodeForUsage(root map[string]any, usage string) (map[string]any, error) {
	matches, err := keyNodesForUsage(root, usage)
	if err != nil {
		return nil, err
	}
	if len(matches) != 1 {
		return nil, fmt.Errorf("trust root does not contain one %s key", usage)
	}
	return matches[0], nil
}

func keyNodeByID(root map[string]any, keyID string) (map[string]any, error) {
	var match map[string]any
	for _, item := range root["keys"].([]any) {
		key := item.(map[string]any)
		if key["key_id"] == keyID {
			if match != nil {
				return nil, fmt.Errorf("key %q is not unique in trust root", keyID)
			}
			match = key
		}
	}
	if match == nil {
		return nil, fmt.Errorf("key %q is absent from trust root", keyID)
	}
	return match, nil
}

func rootKeyView(node map[string]any) RootKey {
	principal := node["principal"].(map[string]any)
	return RootKey{
		KeyID: node["key_id"].(string), Usage: node["usage"].(string),
		PublicKeyBase64URL: node["public_key_base64url"].(string),
		AuthorityDomain:    principal["authority_domain"].(string),
		PrincipalID:        principal["principal_id"].(string),
		PrincipalType:      principal["principal_type"].(string),
	}
}

// Identity returns the caller-supplied self digest and epoch. Neither is an
// external root pin.
func (root *TrustRoot) Identity() (string, int64) {
	return root.document["root_sha256"].(string), root.document["trust_epoch"].(int64)
}

// ResolveKey returns one detached root key by ID.
func (root *TrustRoot) ResolveKey(keyID string) (RootKey, error) {
	node, err := keyNodeByID(root.document, keyID)
	if err != nil {
		return RootKey{}, err
	}
	return rootKeyView(node), nil
}

File: harness/authenticated_adr_approval_contract/authority.py
"""Signature profile and externally pinned independent trust-root shapes."""

from __future__ import annotations

from typing import Any

from .canonical import ContractError, bounded_canonical_json, self_digest
from .constants import (
    CANONICALIZATION,
    KEY_USAGES,
    MAX_PROFILE_BYTES,
    MAX_ROOT_BYTES,
    PROFILE_ID,
    SIGNATURE_PROFILE_API,
    SIGNATURE_PROFILE_DOMAIN,
    SIGNATURE_PROFILE_ID,
    TRUST_ROOT_API,
    TRUST_ROOT_DOMAIN,
)
from .shape import (fixed_base64url, integer, principal_identity, require_keys,
                    sha256, sorted_unique_nodes, text, validate_principal)

SIGNATURE_PROFILE_FIELDS = {
    "algorithm", "api_version", "canonicalization", "digest_algorithm", "kind",
    "message_preimage", "profile_id", "profile_sha256", "public_key_encoding",
    "signature_encoding",
}
TRUST_ROOT_FIELDS = {
    "api_version", "canonicalization", "keys", "kind", "profile_id", "root_sha256",
    "signature_profile_sha256", "trust_domain", "trust_epoch",
}
KEY_FIELDS = {"key_id", "principal", "public_key_base64url", "usage"}
PROFILE_CONSTANTS = {
    "algorithm": "Ed25519",
    "api_version": SIGNATURE_PROFILE_API,
    "canonicalization": CANONICALIZATION,
    "digest_algorithm": "SHA-256",
    "kind": "SignatureProfile",
    "message_preimage": "domain_separator_utf8_nul_then_raw_32_byte_sha256_digest",
    "profile_id": SIGNATURE_PROFILE_ID,
    "public_key_encoding": "base64url_unpadded_32_bytes",
    "signature_encoding": "base64url_unpadded_64_bytes",
}


def signature_profile_sha256(value: dict[str, Any]) -> str:
    return self_digest(SIGNATURE_PROFILE_DOMAIN, value, ("profile_sha256",),
                       MAX_PROFILE_BYTES, "SignatureProfile")


def validate_signature_profile(value: Any) -> dict[str, Any]:
    node = require_keys(value, "SignatureProfile", SIGNATURE_PROFILE_FIELDS)
    bounded_canonical_json(node, MAX_PROFILE_BYTES, "SignatureProfile")
    for field, expected in PROFILE_CONSTANTS.items():
        if node[field] != expected:
            raise ContractError(f"SignatureProfile.{field} drifted from v1")
    sha256(node["profile_sha256"], "SignatureProfile.profile_sha256")
    if node["profile_sha256"] != signature_profile_sha256(node):
        raise ContractError("SignatureProfile self digest does not match")
    return node


def trust_root_sha256(value: dict[str, Any]) -> str:
    return self_digest(TRUST_ROOT_DOMAIN, value, ("root_sha256",), MAX_ROOT_BYTES,
                       "ArchitectureDecisionApprovalTrustRoot")


def validate_trust_root(value: Any, profile_hash: str) -> dict[str, Any]:
    label = "ArchitectureDecisionApprovalTrustRoot"
    node = require_keys(value, label, TRUST_ROOT_FIELDS)
    bounded_canonical_json(node, MAX_ROOT_BYTES, label)
    expected = (TRUST_ROOT_API, CANONICALIZATION, label, PROFILE_ID)
    actual = (node["api_version"], node["canonicalization"], node["kind"],
              node["profile_id"])
    if actual != expected:
        raise ContractError(f"{label} envelope drifted from v1")
    text(node["trust_domain"], f"{label}.trust_domain")
    integer(node["trust_epoch"], f"{label}.trust_epoch", 1, 2**63 - 1)
    if sha256(node["signature_profile_sha256"], "signature_profile_sha256") != profile_hash:
        raise ContractError(f"{label} does not bind the SignatureProfile")
    _validate_keys(node["keys"])
    sha256(node["root_sha256"], f"{label}.root_sha256")
    if node["root_sha256"] != trust_root_sha256(node):
        raise ContractError(f"{label} self digest does not match")
    return node


def _validate_keys(value: Any) -> None:
    keys = sorted_unique_nodes(value, "trust_root.keys", 6, 20)
    for index, key in enumerate(keys):
        _validate_key(key, f"trust_root.keys[{index}]")
    usages = [key["usage"] for key in keys]
    expected_counts = {
        "approval_authorization_state_sign": 1,
        "approval_policy_sign": 1,
        "approval_request_auth": 1,
        "approval_revocation_sign": 1,
    }
    if any(usages.count(usage) != count for usage, count in expected_counts.items()):
        raise ContractError("trust root service/request key usage counts drifted from v1")
    if not 2 <= usages.count("architecture_approval_sign") <= 16:
        raise ContractError("trust root requires 2..16 architecture approval keys")
    identities = [principal_identity(key["principal"]) for key in keys]
    if (len({key["key_id"] for key in keys}) != len(keys) or
            len(set(identities)) != len(keys) or
            len({key["public_key_base64url"] for key in keys}) != len(keys)):
        raise ContractError("root key IDs, principals, and public keys must be pairwise distinct")


def _validate_key(value: Any, label: str) -> None:
    node = require_keys(value, label, KEY_FIELDS)
    text(node["key_id"], f"{label}.key_id")
    if node["usage"] not in KEY_USAGES:
        raise ContractError(f"{label}.usage is unsupported")
    allowed = (("human", "operator") if node["usage"] == "architecture_approval_sign"
               else ("agent", "human", "operator", "service")
               if node["usage"] == "approval_request_auth" else ("service",))
    validate_principal(node["principal"], f"{label}.principal", allowed)
    fixed_base64url(node["public_key_base64url"], f"{label}.public_key_base64url", 32)


def keys_for_usage(root: dict[str, Any], usage: str) -> list[dict[str, Any]]:
    if usage not in KEY_USAGES:
        raise ContractError(f"unsupported trust-root key usage {usage!r}")
    return [key for key in root["keys"] if key["usage"] == usage]


def key_for_usage(root: dict[str, Any], usage: str) -> dict[str, Any]:
    matches = keys_for_usage(root, usage)
    if len(matches) != 1:
        raise ContractError(f"trust root does not contain one {usage} key")
    return matches[0]


def key_by_id(root: dict[str, Any], key_id: str) -> dict[str, Any]:
    matches = [key for key in root["keys"] if key["key_id"] == key_id]
    if len(matches) != 1:
        raise ContractError(f"key {key_id!r} is not unique in trust root")
    return matches[0]
